In [11]:
import pandas as pd
import numpy as np

# Load the dataset
df = pd.read_csv("/content/adult (1).csv")

# Inspect the dataset
print("Dataset shape:", df.shape)

print("\nDataset information:")
df.info()

print("\nFirst 5 rows:")
display(df.head())

print("\nMissing values BEFORE replacing '?':")
print(df.isnull().sum())

Dataset shape: (32561, 15)

Dataset information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education.num   32561 non-null  int64 
 5   marital.status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital.gain    32561 non-null  int64 
 11  capital.loss    32561 non-null  int64 
 12  hours.per.week  32561 non-null  int64 
 13  native.country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB

First 5 rows:


,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,?,77053,HS-grad,9,Widowed,?,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,?,186061,Some-college,10,Widowed,?,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K



Missing values BEFORE replacing '?':
age               0
workclass         0
fnlwgt            0
education         0
education.num     0
marital.status    0
occupation        0
relationship      0
race              0
sex               0
capital.gain      0
capital.loss      0
hours.per.week    0
native.country    0
income            0
dtype: int64


In [12]:
# Replace " ?" with NaN
df.replace(" ?", np.nan, inplace=True)

# Check missing values again
print("Missing values AFTER replacing '?':")
print(df.isnull().sum())

Missing values AFTER replacing '?':
age               0
workclass         0
fnlwgt            0
education         0
education.num     0
marital.status    0
occupation        0
relationship      0
race              0
sex               0
capital.gain      0
capital.loss      0
hours.per.week    0
native.country    0
income            0
dtype: int64


In [13]:
# Calculate missing value count
missing_count = df.isnull().sum()

# Calculate missing value percentage
missing_percentage = (df.isnull().mean() * 100)

# Combine into a DataFrame
missing_summary = pd.DataFrame({
    "Missing Count": missing_count,
    "Missing Percentage": missing_percentage
})

# Sort by missing count descending
missing_summary = missing_summary.sort_values(
    by="Missing Count",
    ascending=False
)

display(missing_summary)

,Missing Count,Missing Percentage
age,0,0.0
workclass,0,0.0
fnlwgt,0,0.0
education,0,0.0
education.num,0,0.0
marital.status,0,0.0
occupation,0,0.0
relationship,0,0.0
race,0,0.0
sex,0,0.0


In [14]:
# Fill missing categorical values with "Unknown"
df["workclass"] = df["workclass"].fillna("Unknown")
df["occupation"] = df["occupation"].fillna("Unknown")

# Also handle native-country
df["native.country"] = df["native.country"].fillna("Unknown")

# Check remaining missing values
print("Missing values after categorical imputation:")
print(df.isnull().sum())

Missing values after categorical imputation:
age               0
workclass         0
fnlwgt            0
education         0
education.num     0
marital.status    0
occupation        0
relationship      0
race              0
sex               0
capital.gain      0
capital.loss      0
hours.per.week    0
native.country    0
income            0
dtype: int64


In [15]:
# Find numeric columns with missing values
numeric_missing = df.select_dtypes(include=np.number).isnull().sum()

print("Missing values in numeric columns:")
print(numeric_missing)

Missing values in numeric columns:
age               0
fnlwgt            0
education.num     0
capital.gain      0
capital.loss      0
hours.per.week    0
dtype: int64


In [16]:
# Impute missing numeric values with the median
numeric_columns = df.select_dtypes(include=np.number).columns

for col in numeric_columns:
    if df[col].isnull().sum() > 0:
        df[col] = df[col].fillna(df[col].median())

print("Remaining missing values:")
print(df.isnull().sum())

Remaining missing values:
age               0
workclass         0
fnlwgt            0
education         0
education.num     0
marital.status    0
occupation        0
relationship      0
race              0
sex               0
capital.gain      0
capital.loss      0
hours.per.week    0
native.country    0
income            0
dtype: int64


In [46]:
# Count exact duplicate rows
exact_duplicates = df.duplicated().sum()

print("Number of exact duplicate rows:", exact_duplicates)

# Remove exact duplicate rows
df.drop_duplicates(inplace=True)

print("Shape after removing exact duplicates:", df.shape)

Number of exact duplicate rows: 24
Shape after removing exact duplicates: (32537, 15)


In [45]:
# Check column names
print(df.columns.tolist())

['age', 'workclass', 'fnlwgt', 'education', 'education.num', 'marital.status', 'occupation', 'relationship', 'race', 'sex', 'capital.gain', 'capital.loss', 'hours.per.week', 'native.country', 'income']


In [19]:
# Remove the target column for duplicate checking
features_only = df.drop(columns=["income"])

# Count duplicate rows based only on features
feature_duplicates = features_only.duplicated().sum()

print("Duplicates ignoring income/target column:", feature_duplicates)

Duplicates ignoring income/target column: 25


In [21]:
print("Unique education values:")
print(df["education"].unique())

print("\nUnique marital-status values:")
print(df["marital.status"].unique())

print("\nUnique native-country values:")
print(df["native.country"].unique())

Unique education values:
['HS-grad' 'Some-college' '7th-8th' '10th' 'Doctorate' 'Prof-school'
 'Bachelors' 'Masters' '11th' 'Assoc-acdm' 'Assoc-voc' '1st-4th' '5th-6th'
 '12th' '9th' 'Preschool']

Unique marital-status values:
['Widowed' 'Divorced' 'Separated' 'Never-married' 'Married-civ-spouse'
 'Married-spouse-absent' 'Married-AF-spouse']

Unique native-country values:
['United-States' '?' 'Mexico' 'Greece' 'Vietnam' 'China' 'Taiwan' 'India'
 'Philippines' 'Trinadad&Tobago' 'Canada' 'South' 'Holand-Netherlands'
 'Puerto-Rico' 'Poland' 'Iran' 'England' 'Germany' 'Italy' 'Japan' 'Hong'
 'Honduras' 'Cuba' 'Ireland' 'Cambodia' 'Peru' 'Nicaragua'
 'Dominican-Republic' 'Haiti' 'El-Salvador' 'Hungary' 'Columbia'
 'Guatemala' 'Jamaica' 'Ecuador' 'France' 'Yugoslavia' 'Scotland'
 'Portugal' 'Laos' 'Thailand' 'Outlying-US(Guam-USVI-etc)']


In [23]:
# Strip leading/trailing whitespace from all object columns
object_columns = df.select_dtypes(include="object").columns

for col in object_columns:
    df[col] = df[col].str.strip()

# Check unique values again
print("Education:")
print(df["education"].unique())

print("\nMarital Status:")
print(df["marital.status"].unique())

print("\nNative Country:")
print(df["native.country"].unique())

Education:
['HS-grad' 'Some-college' '7th-8th' '10th' 'Doctorate' 'Prof-school'
 'Bachelors' 'Masters' '11th' 'Assoc-acdm' 'Assoc-voc' '1st-4th' '5th-6th'
 '12th' '9th' 'Preschool']

Marital Status:
['Widowed' 'Divorced' 'Separated' 'Never-married' 'Married-civ-spouse'
 'Married-spouse-absent' 'Married-AF-spouse']

Native Country:
['United-States' '?' 'Mexico' 'Greece' 'Vietnam' 'China' 'Taiwan' 'India'
 'Philippines' 'Trinadad&Tobago' 'Canada' 'South' 'Holand-Netherlands'
 'Puerto-Rico' 'Poland' 'Iran' 'England' 'Germany' 'Italy' 'Japan' 'Hong'
 'Honduras' 'Cuba' 'Ireland' 'Cambodia' 'Peru' 'Nicaragua'
 'Dominican-Republic' 'Haiti' 'El-Salvador' 'Hungary' 'Columbia'
 'Guatemala' 'Jamaica' 'Ecuador' 'France' 'Yugoslavia' 'Scotland'
 'Portugal' 'Laos' 'Thailand' 'Outlying-US(Guam-USVI-etc)']


In [24]:
# Descriptive statistics for numeric columns
display(df.describe())

,age,fnlwgt,education.num,capital.gain,capital.loss,hours.per.week
count,32561.000000,3.256100e+04,32561.000000,32561.000000,32561.000000,32561.000000
mean,38.581647,1.897784e+05,10.080679,1077.648844,87.303830,40.437456
std,13.640433,1.055500e+05,2.572720,7385.292085,402.960219,12.347429
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.178270e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.783560e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.370510e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.484705e+06,16.000000,99999.000000,4356.000000,99.000000


In [25]:
# Descriptive statistics for categorical columns
display(df.describe(include="object"))

,workclass,education,marital.status,occupation,relationship,race,sex,native.country,income
count,32561,32561,32561,32561,32561,32561,32561,32561,32561
unique,9,16,7,15,6,5,2,42,2
top,Private,HS-grad,Married-civ-spouse,Prof-specialty,Husband,White,Male,United-States,<=50K
freq,22696,10501,14976,4140,13193,27816,21790,29170,24720


In [28]:
# Calculate mean and median
age_mean = df["age"].mean()
age_median = df["age"].median()

hours_mean = df["hours.per.week"].mean()
hours_median = df["hours.per.week"].median()

print("Age:")
print("Mean:", age_mean)
print("Median:", age_median)

print("\nHours-per-week:")
print("Mean:", hours_mean)
print("Median:", hours_median)

Age:
Mean: 38.58164675532078
Median: 37.0

Hours-per-week:
Mean: 40.437455852092995
Median: 40.0


In [29]:
age_difference = abs(age_mean - age_median)
hours_difference = abs(hours_mean - hours_median)

print(f"Difference between age mean and median: {age_difference:.2f}")
print(f"Difference between hours-per-week mean and median: {hours_difference:.2f}")

if age_difference < 2:
    print("Age: Mean and median are relatively close.")
else:
    print("Age: Mean and median differ noticeably.")

if hours_difference < 2:
    print("Hours-per-week: Mean and median are relatively close.")
else:
    print("Hours-per-week: Mean and median differ noticeably.")

Difference between age mean and median: 1.58
Difference between hours-per-week mean and median: 0.44
Age: Mean and median are relatively close.
Hours-per-week: Mean and median are relatively close.


In [30]:
# Frequency of occupations
occupation_counts = df["occupation"].value_counts()

print("Occupation frequencies:")
display(occupation_counts)

print("\nMost common occupation:")
print(occupation_counts.idxmax())

print("Number of people:")
print(occupation_counts.max())

Occupation frequencies:


,count
occupation,
Prof-specialty,4140
Craft-repair,4099
Exec-managerial,4066
Adm-clerical,3770
Sales,3650
Other-service,3295
Machine-op-inspct,2002
?,1843
Transport-moving,1597



Most common occupation:
Prof-specialty
Number of people:
4140


In [31]:
# Percentage distribution of sex
sex_percentage = df["sex"].value_counts(normalize=True) * 100

print("Sex percentage distribution:")
display(sex_percentage.round(2))

Sex percentage distribution:


,proportion
sex,
Male,66.92
Female,33.08


In [32]:
# Percentage distribution of income
income_percentage = df["income"].value_counts(normalize=True) * 100

print("Income percentage distribution:")
display(income_percentage.round(2))

Income percentage distribution:


,proportion
income,
<=50K,75.92
>50K,24.08


In [33]:
# Determine whether income classes are balanced
income_counts = df["income"].value_counts()

print("Income counts:")
display(income_counts)

ratio = income_counts.min() / income_counts.max()

print(f"Minority-to-majority class ratio: {ratio:.2f}")

if ratio >= 0.8:
    print("The income classes are relatively balanced.")
else:
    print("The income classes are imbalanced.")

Income counts:


,count
income,
<=50K,24720
>50K,7841


Minority-to-majority class ratio: 0.32
The income classes are imbalanced.


In [35]:
# Check which education-num values correspond to each education level
education_mapping = df.groupby("education")["education.num"].unique()

display(education_mapping)

,education.num
education,
10th,[6]
11th,[7]
12th,[8]
1st-4th,[2]
5th-6th,[3]
7th-8th,[4]
9th,[5]
Assoc-acdm,[12]
Assoc-voc,[11]


In [36]:
# Find education levels that have more than one education-num value
inconsistent_education = education_mapping[
    education_mapping.apply(len) > 1
]

print("Inconsistent education mappings:")
display(inconsistent_education)

if len(inconsistent_education) == 0:
    print("No inconsistencies found.")
else:
    print("Inconsistencies were found.")

Inconsistent education mappings:


,education.num
education,


No inconsistencies found.


In [38]:
import sqlite3

# Read the RAW CSV again so SQL works on the original data
raw_df = pd.read_csv("/content/adult (1).csv")

# Create an in-memory SQLite database
connection = sqlite3.connect(":memory:")

# Store CSV data in SQLite table
raw_df.to_sql(
    "adult_income",
    connection,
    if_exists="replace",
    index=False
)

print("CSV successfully loaded into SQLite table: adult_income")

CSV successfully loaded into SQLite table: adult_income


In [39]:
# Display first 5 rows from SQLite
sql_check = pd.read_sql_query(
    "SELECT * FROM adult_income LIMIT 5",
    connection
)

display(sql_check)

,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,?,77053,HS-grad,9,Widowed,?,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,?,186061,Some-college,10,Widowed,?,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K


In [40]:
# SQL query
query = """
SELECT *
FROM adult_income
WHERE age > 30
"""

# Execute SQL query and load result into pandas
sql_df = pd.read_sql_query(query, connection)

print("SQL extraction shape:", sql_df.shape)

display(sql_df.head())

SQL extraction shape: (21989, 15)


,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,?,77053,HS-grad,9,Widowed,?,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,?,186061,Some-college,10,Widowed,?,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K


In [41]:
# Number of rows in original CSV satisfying age > 30
expected_rows = (raw_df["age"] > 30).sum()

print("Expected number of rows where age > 30:", expected_rows)
print("Rows returned by SQL:", len(sql_df))

# Compare columns
print("\nOriginal CSV columns:", len(raw_df.columns))
print("SQL result columns:", len(sql_df.columns))

if len(sql_df) == expected_rows:
    print("\nSQL extraction is correct.")
else:
    print("\nSQL extraction does not match the expected result.")

Expected number of rows where age > 30: 21989
Rows returned by SQL: 21989

Original CSV columns: 15
SQL result columns: 15

SQL extraction is correct.


In [42]:
# Pandas equivalent of the SQL query
pandas_df = raw_df[raw_df["age"] > 30].copy()

print("Pandas filtering shape:", pandas_df.shape)
print("SQL filtering shape:", sql_df.shape)

if pandas_df.shape == sql_df.shape:
    print("\nSQL and pandas extraction have the same shape.")
else:
    print("\nSQL and pandas extraction have different shapes.")

Pandas filtering shape: (21989, 15)
SQL filtering shape: (21989, 15)

SQL and pandas extraction have the same shape.


In [43]:
connection.close()

print("SQLite connection closed.")

SQLite connection closed.


## Task 10 — Data Profiling Summary Report

### Dataset Overview

The dataset consists of **32,561 records and 15 columns** in its original form. After removing exact duplicate records, the cleaned dataset contains **32,537 rows and 15 columns**. This is the **UCI Adult Income dataset**, which contains information about individuals from the 1994 U.S. Census. The variables describe characteristics such as age, education, employment, marital status, occupation, and working hours, with the income category indicating whether an individual's income is **<=50K or >50K**.

### Data Quality Issues Found

Several data quality problems were identified during the profiling process:

- **Missing data:** The dataset originally stored missing entries as `"?"`, which were not recognized as missing by pandas. After converting these values to `NaN`, **occupation had 1,843 missing values, workclass had 1,836, and native-country had 583**.
- **Duplicate records:** There were **24 exact duplicate rows**. When the income column was excluded from the duplicate check, **25 duplicate matches** were identified. The additional matching records had different income labels, so they were retained because the same recorded characteristics can occur for different individuals.
- **Categorical inconsistencies:** Several categorical fields contained unnecessary whitespace around their values. This could cause identical categories to be interpreted as separate values, so whitespace was removed from the categorical columns.

### Key Findings

The analysis showed that the income variable is noticeably **imbalanced**. Approximately **75.9%** of observations belong to the `<=50K` group, while only about **24.1%** belong to the `>50K` group. The sex variable is also unevenly distributed, with approximately **66.9% Male** and **33.1% Female**.

For occupation, **Prof-specialty** is the most frequently occurring category, with **4,136 observations**. Craft-repair and Exec-managerial are also among the more common occupations.

The average age is approximately **38.6 years**, compared with a median of **37 years**. Similarly, the average number of hours worked per week is around **40.4 hours**, while the median is **40 hours**. Since the mean and median values are fairly close for both variables, there does not appear to be a major difference between the center of these distributions, although some mild right skew may be present.

The relationship between `education` and `education-num` was also checked. **No inconsistencies were found**: every education category corresponds to one specific education-num value.

### Cleaning Decisions

The `"?"` placeholders were first converted into proper `NaN` values so that missing data could be correctly identified and handled. Missing values in **workclass** and **occupation** were replaced with an `"Unknown"` category instead of deleting those observations, which preserves the available information in the remaining columns.

For **native-country**, the missing entries were filled using the **mode, United-States**, because the missing proportion was relatively small at approximately **1.8%** and the column is strongly dominated by that country.

The **24 exact duplicate rows** were removed from the cleaned dataset. However, records that were identical in their attributes but differed in the income label were not removed because their different outcomes may represent valid observations.

Finally, leading and trailing spaces were removed from categorical variables to ensure that the same category was not accidentally stored under multiple slightly different strings.